src/result.rs

```diff
@@ -0,0 +1,103 @@
+use serde_json;
+use std;
+use std::any::Any;
+
+#[derive(Debug)]
+pub enum Error {
+    IO(std::io::Error),
+    JSON(serde_json::Error),
+    AddrParse(std::net::AddrParseError),
+    JoinError(Box<Any + Send + 'static>),
+    RecvError(std::sync::mpsc::RecvError),
+    SendError,
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+impl std::convert::From<std::sync::mpsc::RecvError> for Error {
+    fn from(e: std::sync::mpsc::RecvError) -> Error {
+        Error::RecvError(e)
+    }
+}
+impl<T> std::convert::From<std::sync::mpsc::SendError<T>> for Error {
+    fn from(_e: std::sync::mpsc::SendError<T>) -> Error {
+        Error::SendError
+    }
+}
+impl std::convert::From<Box<Any + Send + 'static>> for Error {
+    fn from(e: Box<Any + Send + 'static>) -> Error {
+        Error::JoinError(e)
+    }
+}
+
+impl std::convert::From<std::io::Error> for Error {
+    fn from(e: std::io::Error) -> Error {
+        Error::IO(e)
+    }
+}
+impl std::convert::From<serde_json::Error> for Error {
+    fn from(e: serde_json::Error) -> Error {
+        Error::JSON(e)
+    }
+}
+impl std::convert::From<std::net::AddrParseError> for Error {
+    fn from(e: std::net::AddrParseError) -> Error {
+        Error::AddrParse(e)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use result::Result;
+    use result::Error;
+    use std::net::SocketAddr;
+    use std::sync::mpsc::RecvError;
+    use std::thread;
+    use std::io;
+    use std::io::Write;
+    use serde_json;
+    use std::sync::mpsc::channel;
+
+    fn addr_parse_error() -> Result<SocketAddr> {
+        let r = "12fdfasfsafsadfs".parse()?;
+        return Ok(r);
```

Comment:
These two lines can be reduced to:

```rust
"12fdfasfsafsadfs".parse()
```

Reply:
it doesn't do the conversion
```
  --> src/result.rs:74:9                                                                                                                                      
   |                                                                                                                                                          
73 |     fn join_error() -> Result<()> {                                                                                                                      
   |                        ---------- expected `std::result::Result<(), resul                                                                                
t::Error>` because of return type                                                                                                                             
74 |         thread::spawn(|| panic!("hi")).join()                                                                                                            
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected enum `result::Erro                                                                                
r`, found struct `std::boxed::Box`                                                                                                                            
   |                                                                                                                                                          
   = note: expected type `std::result::Result<(), result::Error>`                                                                                             
              found type `std::result::Result<_, std::boxed::Box<std::any::Any                                                                                
 + std::marker::Send + 'static>>`   
```
```
    fn join_error() -> Result<()> {
        thread::spawn(|| panic!("hi")).join()                                                                                                                 
    }
```
or
```
    fn join_error() -> Result<()> {
        Result::from(thread::spawn(|| panic!("hi")).join())                                                                                                   
    }
```
doesn't work either